### ISSUE TITLE
Distribute Elasticsearch documents across multiple indexes

### ISSUE THREAD

**Issue by ayco-at-naturalis** [Sp2000/colplus-backend#563]
If we slightly partition the documents according to dataset key, it won't be so costly to index / sync a dataset (which requires a costly ES _delete_by_query call). Dataset-confined searches will also become faster.

@mdoering my idea would be to add a "doc_bucket" (or so) column to the dataset table, and a method to the dataset mapper that produces a dataset-to-bucket hashmap. 

The ES code will use the bucket name or number to generate index names, and to direct the searches and syncs

**Comment by mdoering** [comment]
How many indices do you think we should target? Should they all be equal size or better be assigned according to their characteristics (read only datasets vs managed catalogues)

**Comment by ayco-at-naturalis** [comment]
I think I would strive to make them by and large equals size. Intuitively I would say about 4 indices, or a million documents per index.

But of course if there are some big datasets that are read-only we could even use index aliases and (on PG import) swap them in & out without time loss. But maybe not for now.

**Comment by mdoering** [comment]
In GBIF ChecklistBank we have 26 million records right now. So expect about 30 million in the next weeks and up to 40 million by the end of next year. As datasets are very different ranging from 1 record to 6 million records creating equal partitions is largely impossible. Assigning a bucket based on the dataset size would require us to first index it - the snake bites its tail. So best we can do is random shuffling and maybe keeping plazi datasets together as these are all small. Or keep the CoL datasets in a special index

**Comment by ayco-at-naturalis** [comment]
Yes, I would reserve special buckets for the CoL datasets (so that in the future we could swap them in & out using aliases). Really big datasets (5m and up) should maybe also have their own bucket, for the same reason. And the rest random shuffling. So that's round and about 10 million documents per index, which should be peanuts for ES.

**Comment by mdoering** [comment]
How do we know the size of a dataset without importing it before? We know col releases and the draft catalogue are rather large. But to manage the others manually seems problematic. And we cannot simply change the bucket once we have already indexed into it.

How does this deal with rebalancing and adding new indices in the future?

**Comment by ayco-at-naturalis** [comment]
That's why I was hopinh that the PG import would somehow be able to populate the doc_bucket quasi-intelligently. Rebalancing and/or adding new indices is of course going to be a costly thing, unless we're dealing with a totally new dataset and we conclude that all existing indices are full according to some chosen treshold.

**Comment by mdoering** [comment]
Right, we have the normalizer phase first that tells us how large a dataset is.

Actually I dont think we need that really. To keep the balance between indices/partitions we would just need to know the smallest index and assign that to the currently imported dataset (only the first time. If it is reimported it keeps it partition). To know the smallest partition we'd need to sum up all dataset counts which I have as metrics at hand. Doesn't sound too complex.

p1: draft col & names index
p2-5: other datasets
